linux/just_files/just_pre-commit_functions.bsh

```diff
@@ -0,0 +1,294 @@
+#
+# .. command:: pre-commit_init
+#
+# Initialize pre-commit for the first time in a new project. This should only be run once and then the resulting files saved to the repo
+#
+# .. command:: pre-commit_setup
+#
+# Sets up the pre-commit virtual environment. If ``pip`` is not found, it will use conda to install a version of python in :envvar:`${JUST_PROJECT_PREFIX}_PRECOMMIT_PYTHON_DIR`. This function is also responsible for the initial installing of ``pip-tools`` into the virtual environment.
+#
+# .. command:: pre-commit_sync
+#
+# Updates the pre-commit's virtual environment (using ``pip-sync``) to have the correct versions specified in :envvar:`${JUST_PROJECT_PREFIX}_PRECOMMIT_DIR`/``requirements.txt``.
+#
+# .. command:: pre-commit_compile
+#
+# Compiles the requirements file for pre-commit, updating :envvar:`${JUST_PROJECT_PREFIX}_PRECOMMIT_DIR`/``requirements.txt``.
+#
+# .. command:: pre-commit_enable
+#
+# Enables pre-commit hooks (via ``pre-commit install``) according to :envvar:`${JUST_PROJECT_PREFIX}_PRECOMMIT_HOOKS`
+#
+# .. command:: pre-commit_disable
+#
+# Disable all pre-commit hooks (via ``pre-commit uninstall``)
+#
+# .. command:: pre-commit_purge
+#
+# Purge the pre-commit virtual environment and remove all hooks.
+#
+# --
+# CI
+# --
+#
+# You can use the following in gitlab CI to run pre-commit checks:
+#
+# .. code:: yaml
+#
+#    pre-commit:
+#      image: alpine:3.19
+#      stage: pre-commit
+#      before_script:
+#        - apk --no-cache add bash
+#      script:
+#        - bash -c "${CI_PROJECT_DIR}/external/vsi_common/linux/ci/pre-commit.bsh"
+#**
+
+function pre-commit::_get_hook_files()
+{
+  local git_dir=$(${GIT} rev-parse --git-dir)
+  local hook_index=0
+  hook_files=("${git_dir}"/hooks/*)
+  for hook_index in ${!hook_files[@]}; do
+    # Remove unexpanded * files            and filenames with extensions, i.e. .sample files
+    if [ ! -f "${hook_files[hook_index]}" ] || [[ $(basename "${hook_files[hook_index]}") =~ \. ]]; then
+      unset hook_files[hook_index]
+    fi
+  done
+}
+
+
+function pre-commit()
+{
+  # this Does't work
+  # PRE_COMMIT_HOME=${precommit_dir}/cache
+
+  local subcommand=()
+  if (( ${#} )); then
+    case "${1}" in
+      autoupdate|init-templatedir|install|install-hooks|migrate-config|run|try-repo|uninstall)
+        # The commands are executed in some weird .cache dir, relative path to requirements will never work
+        sed 's|\${precommit_dir}|'"${precommit_dir}"'|g' "${precommit_dir}/config.yml.in" > "${precommit_dir}/config.yml"
+        subcommand=("${1}" -c "${precommit_dir}/config.yml")
+        shift 1
+        ;;
+    esac
+  fi
+
+  local JUST_IGNORE_EXIT_CODES=1
+  ${DRYRUN} "$(vsi::python::virtualenv_bin "${python_dir}" pre-commit)" \
+    ${subcommand[@]+"${subcommand[@]}"} ${@+"${@}"}
+
+  if (( ${#subcommand[@]} )) && [[ ${subcommand[0]} = install* ]]; then
+    local hook hook_files
```

Comment:
I thought there was a problem with some bash when you create multiple variables on the same line

Reply:
```
# local foo="$(bar)" || rv="${?}" does not work as one line in bash 4.2 or older. Use two lines
```

This should be ok